plugins/modules/redfish_firmware_rollback.py

```diff
@@ -0,0 +1,345 @@
+
+
+UPDATE_SERVICE = "UpdateService"
+SYSTEM_RESOURCE_ID = "System.Embedded.1"
+NO_COMPONENTS = "There were no firmware components to rollback."
+COMP_FAIL = "There were no matches found to the firmware component with a '{0}'."
+CHANGES_FOUND = "Changes found to be applied."
+NO_CHANGES_FOUND = "No changes found to be applied."
+NOT_SUPPORTED = "The target firmware version does not support the firmware rollback."
+COMPLETED_ERROR = "The job for firmware rollback has been completed with error(s)."
+SCHEDULED_ERROR = "The job for firmware rollback has been scheduled with error(s)."
+ROLLBACK_SUCCESS = "Successfully completed the job for firmware rollback."
+ROLLBACK_SCHEDULED = "Successfully scheduled the job for firmware rollback."
+ROLLBACK_FAILED = "Failed to complete the job for firmware rollback."
+REBOOT_FAIL = "Failed to reboot the server."
+REBOOT_COMP = ["Integrated Dell Remote Access Controller"]
+
+
+def get_rollback_preview_target(redfish_obj, module):
+    action_resp = redfish_obj.invoke_request("GET", "{0}{1}".format(redfish_obj.root_uri, UPDATE_SERVICE))
+    action_attr = action_resp.json_data["Actions"]
+    update_uri = None
+    if "#UpdateService.SimpleUpdate" in action_attr:
+        update_service = action_attr.get("#UpdateService.SimpleUpdate")
+        if 'target' in update_service:
+            update_uri = update_service.get('target')
+        else:
+            module.fail_json(msg=NOT_SUPPORTED)
+    inventory_uri = action_resp.json_data.get('FirmwareInventory').get('@odata.id')
+    inventory_uri_resp = redfish_obj.invoke_request("GET", "{0}{1}".format(inventory_uri, "?$expand=*($levels=1)"),
+                                                    api_timeout=60)
+    previous_component = list(filter(lambda d: d["Id"].startswith("Previous"), inventory_uri_resp.json_data["Members"]))
+    if not previous_component:
+        module.fail_json(msg=NO_COMPONENTS)
+    component_name = module.params["name"]
+    component_compile = re.compile(r"^{0}$".format(component_name))
+    prev_uri, reboot_uri = {}, []
+    for each in previous_component:
+        available_comp = each["Name"]
+        available_name = re.match(component_compile, available_comp)
+        if not available_name:
+            continue
+        if available_name.group() in REBOOT_COMP:
+            reboot_uri.append(each["@odata.id"])
+            continue
+        prev_uri[each["Version"]] = each["@odata.id"]
+    if module.check_mode and (prev_uri or reboot_uri):
+        module.exit_json(msg=CHANGES_FOUND, changed=True)
+    elif module.check_mode and (not prev_uri and not reboot_uri):
+        module.exit_json(msg=NO_CHANGES_FOUND)
+    elif not module.check_mode and (not prev_uri and not reboot_uri):
+        module.exit_json(msg=COMP_FAIL.format(component_name))
+    return list(prev_uri.values()), reboot_uri, update_uri
+
+
+def get_job_status(redfish_obj, module, job_ids, job_wait=True):
```

Comment:
same here, can be moved to utils

Reply:
`get_job_status` - function internally calls `wait_for_redfish_job_complete` this function is in utils.

Reply:
can't we use this function in another module to get job status, if we can , shouldn't we move to utils ...

Reply:
We can use it, here it returns `failed_count` that is specific to this module use case.